src/services/persistent-storage.service.spec.ts

```diff
@@ -0,0 +1,106 @@
+import { PersistentStorage } from "./persistent-storage.service";
+import { mock, mockReset } from "jest-mock-extended";
+import { defer } from "../utils/HelperFunctions";
+
+describe("PersistentStorage Unit Tests", () => {
+  let storage: Storage;
+  let persistentStorage: PersistentStorage;
+
+  beforeAll(() => {
+    storage = mock<Storage>();
+    persistentStorage = new PersistentStorage("testing", storage);
+  });
+
+  afterEach(async () => {
+    // Reset the mock after the JS stack has cleared
+    await defer(() => {});
+    mockReset(storage);
+    await persistentStorage.clear();
+    await defer(() => {});
+  });
+
+  it("should exist", () => {
+    expect(PersistentStorage).toBeTruthy();
+  });
+
+  it("should instanciate as a singleton", () => {
+    const persistentStorage2 = new PersistentStorage("testing", storage);
+    expect(persistentStorage).toBe(persistentStorage2);
+  });
+
+  it('should init properly', async () => {
+    const initStorage = mock<Storage>({
+      getItem: jest.fn().mockImplementation(() => "testValue"),
+      length: 1,
+      key: jest.fn().mockImplementation(() => "inittestKey"),
+    });
+
+    const initPersistantStorage = new PersistentStorage("init", initStorage);
+
+    const value = await initPersistantStorage.getItem("testKey");
+    expect(value).toBe("testValue");
+    expect(initStorage.getItem).toHaveBeenCalledWith("inittestKey");
+    expect(initPersistantStorage.length).toBe(1);
+
+  });
+
+  it("should set and get a value and sync to localstorage", async () => {
+    await persistentStorage.setItem("test", "test");
+    const setValue = await persistentStorage.getItem("test");
+    expect(setValue).toBe("test");
+  });
+
+  it("should defer setting the value to localstorage until after the JS stack has cleared (performance first)", async () => {
+    await persistentStorage.setItem("test", "test");
+    expect(storage.setItem).toBeCalledTimes(0);
+    await defer(() => {});
+    expect(storage.setItem).toBeCalledTimes(1);
```

Comment:
Ideally, we want something like a `waitFor` or as mentioned in another comment, resolve `setItem` once the internal `defer` is called. As it stands, these tests are too concerned with how it works internally. We just care that it writes to the underlying storage